packages/chatbot-server-mongodb-public/src/tools/fetchPage.ts

```diff
@@ -0,0 +1,111 @@
+import { z } from "zod";
+import { FindContentFunc, MongoDbPageStore, Reference } from "mongodb-rag-core";
+import { normalizeUrl } from "mongodb-rag-core/dataSources";
+import { wrapTraced } from "mongodb-rag-core/braintrust";
+import { Tool, tool, ToolExecutionOptions } from "mongodb-rag-core/aiSdk";
+import { addReferenceSourceType } from "../processors/makeMongoDbReferences";
+
+export const FETCH_PAGE_TOOL_NAME = "fetch_page";
+
+export const MongoDbFetchPageToolArgsSchema = z.object({
+  pageUrl: z.string().describe("URL of page to fetch"),
+  query: z.string().describe("Search query"),
+});
+
+export type MongoDbFetchPageToolArgs = z.infer<
+  typeof MongoDbFetchPageToolArgsSchema
+>;
+
+export type FetchPageToolResult = {
+  url: string;
+  text: string;
+  reference?: Reference;
+};
+
+export type FetchPageTool = Tool<
+  typeof MongoDbFetchPageToolArgsSchema,
+  FetchPageToolResult
+> & {
+  execute: (
+    args: MongoDbFetchPageToolArgs,
+    options: ToolExecutionOptions
+  ) => PromiseLike<FetchPageToolResult>;
+};
+
+export function makeFetchPageTool(
+  loadPage: MongoDbPageStore["loadPage"],
+  findContent: FindContentFunc
+): FetchPageTool {
+  return tool({
+    parameters: MongoDbFetchPageToolArgsSchema,
+    description: "Fetch all content for a specific URL",
+    experimental_toToolResultContent(result) {
+      return [{ type: "text", text: result.text }];
+    },
+    execute: wrapTraced(
+      async function (
+        args: MongoDbFetchPageToolArgs,
+        _options: ToolExecutionOptions
+      ): Promise<FetchPageToolResult> {
+        // TODO Comment in once ingestion is normalized too
+        const normalizedUrl = args.pageUrl; //normalizeUrl(args.pageUrl);
+        const page = await loadPage({
+          urls: [normalizedUrl],
+          query: {
+            action: {
+              $ne: "deleted",
+            },
+          },
+        });
+
+        let text: string;
+        let reference: Reference | undefined;
+        if (page === null) {
+          // Fall back - no page for this URL
+          text = "{fallback_to_search}";
+        } else if (page.body.length < 150000) {
```

Comment:
can u put the 150000 magic number as argument in the makeFetchPageTool constructor?

generally good for magic numbers to be configurable